phantazm-zombies/src/main/java/com/github/phantazmnetwork/zombies/equipment/gun/Gun.java

```diff
@@ -0,0 +1,185 @@
+package com.github.phantazmnetwork.zombies.equipment.gun;
+
+import com.github.phantazmnetwork.api.inventory.CachedInventoryObject;
+import com.github.phantazmnetwork.api.player.PlayerView;
+import com.github.phantazmnetwork.mob.MobStore;
+import com.github.phantazmnetwork.zombies.equipment.Equipment;
+import com.github.phantazmnetwork.zombies.equipment.Upgradable;
+import com.github.phantazmnetwork.zombies.equipment.gun.effect.GunEffect;
+import com.github.phantazmnetwork.zombies.equipment.gun.visual.GunStackMapper;
+import net.minestom.server.coordinate.Pos;
+import net.minestom.server.item.ItemStack;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashSet;
+import java.util.Objects;
+import java.util.function.Consumer;
+
+public class Gun extends CachedInventoryObject implements Equipment, Upgradable {
+
+    private final PlayerView playerView;
+
+    private final GunModel model;
+
+    private GunState state;
+
+    private int level;
+
+    private boolean reloadComplete = false;
+
+    public Gun(@NotNull PlayerView playerView, @NotNull GunModel model, @NotNull MobStore store) {
+        this.playerView = Objects.requireNonNull(playerView, "playerView");
+        this.model = Objects.requireNonNull(model, "model");
+        this.level = 0;
+
+        GunStats stats = getLevel().stats();
+        this.state = new GunState(stats.shootSpeed(), stats.reloadSpeed(), stats.maxAmmo(), stats.maxClip(),
+                false, 0);
+    }
+
+    public void shoot() {
+        if (getLevel().shootTester().shouldShoot(state)) {
+            modifyState(state -> state.setQueuedShots(getLevel().stats().shots() - 1));
+            internalShoot();
+        }
+    }
+
+    protected void internalShoot() {
+        modifyState(builder -> {
+            builder.setTicksSinceLastShot(0L);
+            builder.setAmmo(builder.getAmmo() - 1);
+            builder.setClip(builder.getClip() - 1);
+        });
+        if (state.clip() == 0) {
+            if (state.ammo() > 0) {
+                reload();
+            } else {
+                for (GunEffect effect : getLevel().emptyClipEffects()) {
+                    effect.accept(state);
+                }
+            }
+        }
+
+        playerView.getPlayer().ifPresent(player -> {
+            Pos start = player.getPosition().add(0, player.getEyeHeight(), 0);
+            getLevel().firer().fire(state, start, new HashSet<>());
```

Comment:
reminder to self, ignore this, might delete after reading `fire`'s signature: seems weird that it's being passed an empty hashset

Reply:
okay this should be Collections#emptyList()

Reply:
the point is for the collection to be mutable so it can be updated over time

Reply:
but if it's empty your implementations seem to just return

Reply:
wdym

Reply:
nvm i misread, this is fine